File: gitdot-web/app/dal/webhook.ts
import "server-only";

import {
  type CreateWebhookRequest,
  type UpdateWebhookRequest,
  WebhookResource,
} from "gitdot-api";
import { z } from "zod";
import {
  authDelete,
  authFetch,
  authPatch,
  authPost,
  GITDOT_SERVER_URL,
  handleResponse,
} from "./util";

export async function createWebhook(
  owner: string,
  repo: string,
  request: CreateWebhookRequest,
): Promise<WebhookResource | null> {
  const response = await authPost(
    `${GITDOT_SERVER_URL}/repository/${encodeURIComponent(owner)}/${encodeURIComponent(repo)}/webhook`,
    request,
  );

  return await handleResponse(response, WebhookResource);
}

export async function listWebhooks(
  owner: string,
  repo: string,
): Promise<WebhookResource[] | null> {
  const response = await authFetch(
    `${GITDOT_SERVER_URL}/repository/${encodeURIComponent(owner)}/${encodeURIComponent(repo)}/webhooks`,
  );

  return await handleResponse(response, z.array(WebhookResource));
}

export async function updateWebhook(
  owner: string,
  repo: string,
  webhookId: string,
  request: UpdateWebhookRequest,
): Promise<WebhookResource | null> {
  const response = await authPatch(
    `${GITDOT_SERVER_URL}/repository/${encodeURIComponent(owner)}/${encodeURIComponent(repo)}/webhook/${webhookId}`,
    request,
  );

  return await handleResponse(response, WebhookResource);
}

export async function deleteWebhook(
  owner: string,
  repo: string,
  webhookId: string,
): Promise<void> {
  await authDelete(
    `${GITDOT_SERVER_URL}/repository/${encodeURIComponent(owner)}/${encodeURIComponent(repo)}/webhook/${webhookId}`,
  );
}

File: gitdot-server/src/handler/webhook.rs
mod create_webhook;
mod delete_webhook;
mod get_webhook;
mod list_webhooks;
mod update_webhook;

use axum::{
    Router,
    routing::{get, post},
};

use crate::app::AppState;

use create_webhook::create_webhook;
use delete_webhook::delete_webhook;
use get_webhook::get_webhook;
use list_webhooks::list_webhooks;
use update_webhook::update_webhook;

pub fn create_webhook_router() -> Router<AppState> {
    Router::new()
        .route("/repository/{owner}/{repo}/webhook", post(create_webhook))
        .route("/repository/{owner}/{repo}/webhooks", get(list_webhooks))
        .route(
            "/repository/{owner}/{repo}/webhook/{webhook_id}",
            get(get_webhook)
                .patch(update_webhook)
                .delete(delete_webhook),
        )
}
